src/software/firmware/srcs/pressure_controller.cpp

```diff
@@ -346,11 +351,16 @@ void PressureController::safeguardPressionCrete(uint16_t p_centiSec) {
 
 void PressureController::safeguardPressionPlateau(uint16_t p_centiSec) {
     if (m_subPhase == CycleSubPhases::HOLD_INSPIRATION) {
-        if (m_pressure >= m_maxPlateauPressureCommand) {
-            // TODO vérifier avec médical si on doit délester la pression
-            // TODO alarme franchissement plateau haut
+        if (m_pressure < ALARM_THRESHOLD_PLATEAU_UNDER_2_CMH2O) {
+            m_alarmController.detectedAlarm(24u, m_cycleNb);
         } else {
-            plateau();
+            m_alarmController.notDetectedAlarm(24u);
+        }
+
+        if (m_pressure > ALARM_THRESHOLD_PLATEAU_ABOVE_80_CMH2O) {
+            m_alarmController.detectedAlarm(17u, m_cycleNb);
+        } else {
+            m_alarmController.notDetectedAlarm(17u);
```

Comment:
As I understood, safeguards methods are called every 10ms and checks the corresponding m_subPhase. That means safeguards methods can be called several times during the same cycle.
Now if an alarm needs more than one cycle to be triggered, there is here the possibility that the safeguard method is called with a pressure lower than the threshold which lead to a notDetectedAlarm() call --> you lose the alarm, right? 

I guess notDetectedAlarm() methods should consider too the m_cycleNb to prevent from erasing alarm detection during the same cycle.

Reply:
Very good point 🤔 I think the spec is a bit leaky on this point. I will ask to the right people in the team that might have a strong answer on this. As of now, I understand it as "if PEEP command is reached at the end of the cycle, that's OK".

Reply:
Hello,

I know that this topic PR is closed now but what about the alarms which needs more than 1 cycle to be triggered, would they be considered?

Reply:
This is handled individually by each alarm object which stores a number of times in a row it was detected (in `m_detectionNumber`); see https://github.com/covid-response-projects/covid-respirator/blob/master/src/software/firmware/srcs/alarm.cpp#L38

Reply:
Well, I have seen that it is ceiled to the maximum trigger. My point here is the following, take 

> RCM-SW-19 (The device shall embed a low priority alarm 24 when the Pressure < 2cm H2O at the second cycle (Patient disconnection)

and have a look at the code in "pressure_controller.cpp":
```
void PressureController::safeguardPlateau(uint16_t p_centiSec) {
    if (m_subPhase == CycleSubPhases::HOLD_INSPIRATION) {
        if (m_pressure < ALARM_THRESHOLD_PLATEAU_UNDER_2_CMH2O) {
            m_alarmController->detectedAlarm(RCM_SW_19, m_cycleNb);
        } else {
            m_alarmController->notDetectedAlarm(RCM_SW_19);
        }

        if (m_pressure > ALARM_THRESHOLD_PLATEAU_ABOVE_80_CMH2O) {
            m_alarmController->detectedAlarm(RCM_SW_18, m_cycleNb);
        } else {
            m_alarmController->notDetectedAlarm(RCM_SW_18);
        }
    }
}
```
The safeguards function is called evrey 10ms and regarding requirement 19, you need 2 respiratory cycles to trigger this medium alarm. The safeguards works but it can set/unset several times the alarm during the same cycle (Hold inspiration state is set for several turn I guess as it is called every 10ms). Isn't it a issue for alarm detection that last more than 1 cycle? Because if it is set, then it is and must not be unset after another 10ms, right?

Is it clear enough? I am available on slack if necessary to discuss it.
Thomas

Reply:
Alarms are in the process of being fixed in various aspects, see #160. If I understand what you say, I think most of this problem's occurrences are fixed in #160.

But you are right: specs do not tell us when to untrigger alarms. We did what we believed was right, but I am going to check with @cmdm